### PR TITLE
Loging: close window after login if certain query param is set

### PR DESCRIPTION
**Pull request by roccotripaldi**
#### Changes proposed in this Pull Request

* If someone visits the login screen with `close_window_after_login=true` in the URL, attempt to close the window.

#### Testing instructions

* In progress - waiting until D32103-code is merged

### CODE REVIEW

**Comment by sirreal** [comment]
Some thoughts:

- I don't think we need any dedicated state changes to achieve the same effect. There are selectors for query arguments:
  https://github.com/Automattic/wp-calypso/blob/0ecebd1057518f31316043d3dfa844d442e3ede1/client/state/selectors/get-current-query-arguments.js#L18
  https://github.com/Automattic/wp-calypso/blob/0ecebd1057518f31316043d3dfa844d442e3ede1/client/state/selectors/get-initial-query-arguments.js#L18
- I suggested we could do something clever with like [here](https://github.com/Automattic/wp-calypso/pull/35786#issuecomment-524968606), but an explicit close on login parameter might be best.
- I suspect we should make sure there's a parent window before closing. I don't think there's ever a use case for closing a login window without a parent.
- This would benefit from a minimal login app #33642 / @sgomes 

In general, I appreciate more detail in PR descriptions 🙂 This entire project was new to me.

**Comment by sgomes** [comment]
> * This would benefit from a minimal login app #33642 / @sgomes

Thanks for looping me in, @sirreal! Looks like more special behaviour to aware of when the split-out gets underway again.

File: client/blocks/login/index.jsx [hunk 1]
```diff
@@ -129,7 +130,11 @@ class Login extends Component {
 	};
 
 	rebootAfterLogin = () => {
-		const { redirectTo } = this.props;
+		const { redirectTo, closeWindowAfterLogin } = this.props;
+
+		if ( closeWindowAfterLogin ) {
+			window.close();
```
Comment (sgomes) [inline client/blocks/login/index.jsx]: Login does get rendered on the server side, so unguarded statements like this leave me nervous. Can `rebootAfterLogin` ever run on the server? If so, I suggest changing this line to:

```suggestion
			typeof window !== 'undefined' && window.close();
```

Comment (sirreal) [reply client/blocks/login/index.jsx]: 💯 Always guard the `window` 💂

**Comment by sgomes** [comment]
> @sirreal suggested we do
> 
> ```
> if ( window && window.parent && redirect_to === window.parent.document.location ) {
>     window.close()
> }
> ```
> 
> @Automattic/luna could you help us decide which one is the clearest approach here. For further details on what we are trying to accomplish: p7rcWF-14z-p2

Sorry, I hadn't seen this comment before I added my own!

I just want to point out that `if ( window && [...] )` isn't a sufficient guard for the server-side case. That will still produce an error, because of the missing reference. You'll need:

`if ( typeof window !== 'undefined' && [...] )` to avoid the error altogether. #justjsthings

File: client/state/login/test/selectors.js [hunk 2]
```diff
@@ -397,4 +397,6 @@ describe( 'selectors', () => {
 			).to.eql( 'google' );
 		} );
 	} );
+
+	describe( 'getCloseWindowAfterLogin', function() {} );
```
Comment (tyxla) [inline client/state/login/test/selectors.js]: Let's remove that if we won't be adding tests.

File: client/state/login/selectors.js [hunk 3]
```diff
@@ -175,6 +175,16 @@ export const getRedirectToSanitized = state => {
...
 };
 
+/***
+ * Retrieves the last known value of `close_window_after_login` from the query params of a `/login` page.
+ *
+ * @param  {Object}   state  Global state tree
+ * @return {?bool}           True if the window should be closed after login.
+ */
+export const getCloseWindowAfterLogin = state => {
+	return get( state, 'login.closeWindowAfterLogin', null );
```
Comment (tyxla) [inline client/state/login/selectors.js]: Nit: arrays here used to be faster last time I looked.

```suggestion
	return get( state, [ 'login', 'closeWindowAfterLogin' ], null );
```

File: client/state/login/reducer.js [hunk 4]
```diff
@@ -62,6 +62,16 @@ export const isRequesting = createReducer( false, {
 	[ SOCIAL_DISCONNECT_ACCOUNT_REQUEST_SUCCESS ]: () => false,
 } );
 
+export const closeWindowAfterLogin = createReducer( null, {
```
Comment (tyxla) [inline client/state/login/reducer.js]: Do we really need to use `createReducer`? This seems to be a pretty straightforward usage that can be done in a pure function.

File: client/blocks/login/index.jsx [hunk 5]
```diff
@@ -405,6 +411,7 @@ class Login extends Component {
 
 export default connect(
 	state => ( {
+		closeWindowAfterLogin: getCloseWindowAfterLogin( state ),
```
Comment (tyxla) [inline client/blocks/login/index.jsx]: I wonder if we need to introduce a new selector. Could we use the existing `getCurrentQueryArguments` selector instead? Would make this PR simpler.

**Comment by roccotripaldi** [comment]
Abandoning in favor of https://github.com/Automattic/jetpack/pull/13364

**Pull request closed without merging**